### PR TITLE
Story/9417 2

### PR DESCRIPTION
**Pull request by nocturnalastro**

### CODE REVIEW

File: addons/udes_stock/models/stock_move_line.py [hunk 1]
```diff
@@ -1037,3 +1037,23 @@ def _prevent_write_after_done(self):
         done_lines = self.filtered(lambda x: x.state == "done")
         if done_lines:
             raise ValidationError(_("Cannot update move lines that are already 'done'."))
+
+    @api.constrains("result_package_id", "u_result_parent_package_id")
+    def _check_resultant_package_level(self):
+        pack_level = self.result_package_id.u_package_depth
+
+        # Assumes that the current package hierarchy is acceptable
+        if pack_level > max(self.u_result_parent_package_id.children_ids.mapped("u_package_depth")):
```
Comment (unirho) [inline addons/udes_stock/models/stock_move_line.py]: can't we just use the u_package_depth on the u_result_parent_package_id? Why do the max? This is already the max isn't it? Or does the package not exist yet/this is before the children haven't been put in?

File: addons/udes_stock/models/stock_move_line.py [hunk 2]
```diff
@@ -962,22 +1012,75 @@ def _drop_off_criterion_summary(self):
...
+
+        for ml in self:
+            if ml.picking_id.picking_type_id.u_target_storage_format == "product" and (
+                ml.u_result_parent_package_id or ml.result_package_id
+            ):
+                raise ValidationError(_("Pickings stored by product cannot be inside packages."))
+            if (
+                ml.picking_id.picking_type_id.u_target_storage_format
+                in ["package", "pallet_products"]
+                and ml._u_result_parent_package_id
```
Comment (ws-mps) [inline addons/udes_stock/models/stock_move_line.py]: `ml._u_result_parent_package_id` --> `ml.u_result_parent_package_id`

File: addons/udes_stock/models/stock_move_line.py [hunk 3]
```diff
@@ -962,22 +1012,75 @@ def _drop_off_criterion_summary(self):
...
+                (
+                    package_mls.filtered(
+                        lambda package_ml: ml.u_result_parent_package_id
+                        == package_ml.result_package_id
+                    )
+                )
+                or (
+                    package_mls.filtered(
+                        lambda package_ml: ml.u_result_package_id
+                        == package_ml.result_parent_package_id
```
Comment (ws-mps) [inline addons/udes_stock/models/stock_move_line.py]: `package_ml.result_parent_package_id` --> `package_ml.u_result_parent_package_id`

File: addons/udes_stock/models/stock_quant_package.py [hunk 4]
```diff
@@ -15,6 +15,24 @@ class StockQuantPackage(models.Model):
     # different to the internal one.
     u_external_name = fields.Char(string="External Name")
 
+    u_package_depth = fields.Integer(string="Package Depth",
+        help="The maxmium number of package levels within a package hierarchy. I.e. 2 would denote a single level of packages(each with no subpackage) within a parent package",
```
Comment (ws-mps) [inline addons/udes_stock/models/stock_quant_package.py]: Typo: maxmium

Comment (unirho) [reply addons/udes_stock/models/stock_quant_package.py]: resolved

File: addons/udes_stock/models/stock_quant_package.py [hunk 5]
```diff
@@ -15,6 +15,24 @@ class StockQuantPackage(models.Model):
...
+    @api.depends('children_ids.u_package_depth')
+    def _compute_package_depth(self):
+        wh = self.env.user.get_user_warehouse()
+        max_package_depth = wh.u_max_package_depth
+        for package in self:
+            new_package_depth = max(package.children_ids.mapped('u_package_depth')) + 1 if package.children_ids else 1
+            if new_package_depth != package.u_package_depth:
+                package.u_package_depth = new_package_depth
+            if package.u_package_depth > max_package_depth:
+                raise ValidationError("Maximum package depth exceeded.")
```
Comment (nocturnalastro) [inline addons/udes_stock/models/stock_quant_package.py]: translation

File: addons/udes_stock/models/stock_move_line.py [hunk 6]
```diff
@@ -962,22 +1012,76 @@ def _drop_off_criterion_summary(self):
...
+            [
+                ("state", "=", "assigned"),
+                "|",
+                ("result_package_id", "in", package_ids.mapped("id")),
+                ("u_result_parent_package_id", "in", package_ids.mapped("id")),
+            ]
+        )
+
+        for ml in self:
+            storage_format = ml.picking_id.picking_type_id.u_target_storage_format
```
Comment (ws-mps) [inline addons/udes_stock/models/stock_move_line.py]: You can use: `ml.u_picking_type_id` instead of having to get it from the pick.

Comment (unirho) [reply addons/udes_stock/models/stock_move_line.py]: I'll change that! I copied this from somewhere else 🤔 I'll find that and make a note to sort that

**Pull request merged**